### PR TITLE
PR discouted episodic return in ReplayBuffer

### PR DESCRIPTION
**Pull request by le-horizon**
Extracted the code for computing episodic discounted return from my branch.

Cheers,
Le

### CODE REVIEW

File: alf/algorithms/td_loss.py [hunk 1]
```diff
@@ -146,6 +146,19 @@ def compute_td_target(self, info: namedtuple, target_value: torch.Tensor):
                 td_lambda=self._lambda)
             returns = advantages + target_value[:-1]
 
+        disc_ret = ()
+        if hasattr(info, "batch_info") and hasattr(info.batch_info,
```
Comment (emailweixu) [inline alf/algorithms/td_loss.py]: `info` does not have `batch_info` attribute.

File: alf/experience_replayers/replay_buffer.py [hunk 2]
```diff
@@ -291,6 +329,13 @@ def add_batch(self, batch, env_ids=None, blocking=False):
                 self._indexed_pos[(env_ids[epi_first],
                                    self.circular(overwriting_pos[epi_first])
                                    )] = overwriting_pos[epi_first]
+                epi_last, = torch.where(step_types == ds.StepType.LAST)
```
Comment (emailweixu) [inline alf/experience_replayers/replay_buffer.py]: move this into if block

File: alf/experience_replayers/replay_buffer.py [hunk 3]
```diff
@@ -291,6 +329,13 @@ def add_batch(self, batch, env_ids=None, blocking=False):
...
+                if self._keep_episodic_info and self._record_episodic_return:
+                    # Backfill episodic returns for episodes which ended
+                    self._compute_store_episodic_return(
```
Comment (emailweixu) [inline alf/experience_replayers/replay_buffer.py]: no need to do this if epi_last is empty.

File: alf/experience_replayers/replay_buffer.py [hunk 4]
```diff
@@ -30,8 +30,13 @@
...
 
 BatchInfo = namedtuple(
-    "BatchInfo",
-    ["env_ids", "positions", "importance_weights", "replay_buffer"],
+    "BatchInfo", [
+        "env_ids",
+        "positions",
+        "importance_weights",
+        "replay_buffer",
+        "discounted_return",
```
Comment (emailweixu) [inline alf/experience_replayers/replay_buffer.py:38]: I cant find any code that set a value for this.
Also, I feel BatchInfo is not the right place for this value since most of our code does not have easy access to it.
rollout_info may be a better place.

**Review by emailweixu — CHANGES_REQUESTED** [review]

File: alf/algorithms/td_loss.py (hunk 1 above)
Comment (le-horizon) [reply alf/algorithms/td_loss.py]: Populated in Sac and Ddpg.

File: alf/experience_replayers/replay_buffer.py (hunk 4 above)
Comment (le-horizon) [reply alf/experience_replayers/replay_buffer.py:38]: Good catch, populated in get_batch, and then in preprocess_experience to rollout_info, and then train_step to policy_step.info, so td_loss can have access.

File: alf/experience_replayers/replay_buffer.py (hunk 2 above)
Comment (le-horizon) [reply alf/experience_replayers/replay_buffer.py]: done

File: alf/experience_replayers/replay_buffer.py (hunk 3 above)
Comment (le-horizon) [reply alf/experience_replayers/replay_buffer.py]: done

**Review by le-horizon — COMMENTED** [review]
Thanks for the quick review.  Updated the PR for all comments, PTAL.

File: alf/algorithms/ddpg_algorithm.py [hunk 5]
```diff
@@ -359,6 +360,15 @@ def calc_loss(self, info: DdpgInfo):
             priority=priority,
             extra=DdpgLossInfo(critic=critic_loss, actor=actor_loss.extra))
 
+    def preprocess_experience(self, exp, rollout_info, batch_info):
```
Comment (Haichao-Zhang) [inline alf/algorithms/ddpg_algorithm.py]: Is it possible to move this to a centralized place (e.g. Agent), so we do not need to add this for each algorithm?

Comment (le-horizon) [reply alf/algorithms/ddpg_algorithm.py]: Good point.  I've thought about it, but don't have a good way at the moment.  1) We could put it in the OffPolicyAlgorithm parent class.  But subclasses may override it and accidentally skip this preprocessing.  2) Put in Agent.  But then DdpgAlgo standalone won't work.  3) AlgorithmInfo namedtuples are algorithm specific.  Unless we extract some common parts out and put inside say OffPolicyAlgo, there will always be some algorithm specific change involved to utilize discounted_return.

Maybe let's think more about it.

Comment (le-horizon) [reply alf/algorithms/ddpg_algorithm.py]: A cleaner way is to extract the preprocess part into a Algorithm wrapper, containing the ddpg or sac algorithm as a sub-algorithm.  This will make it easy to configure any new algorithm.  Repurposing any existing config file with just command line args would be more difficult, and we'd probably have to create a separate config file for any runs using discounted_return.

@Haichao-Zhang If you are Ok with this, I can handle it perhaps in another PR.

Comment (emailweixu) [reply alf/algorithms/ddpg_algorithm.py]: I think it's ok to put it in Agent. The intention of Agent is an integration holder for different algorithms.

Comment (le-horizon) [reply alf/algorithms/ddpg_algorithm.py]: Sounds good.  Done (only when AlgoInfo has batch_info field).

**Review by Haichao-Zhang — DISMISSED** [review]

**Review by Haichao-Zhang — DISMISSED** [review]

File: alf/algorithms/agent.py [hunk 6]
```diff
@@ -415,6 +415,17 @@ def preprocess_experience(self, root_inputs, rollout_info, batch_info):
 
         exp, rl_info = self._rl_algorithm.preprocess_experience(
             exp, rollout_info.rl, batch_info)
+
+        # Add batch_info into rollout_info to populate discounted_return.
+        if hasattr(rl_info,
+                   "batch_info") and batch_info.discounted_return != ():
```
Comment (emailweixu) [inline alf/algorithms/agent.py]: how about just hasattr(rl_info, "discounted_return") and only pass discounted_return to rl_info to simply code?

File: alf/experience_replayers/replay_buffer.py [hunk 7]
```diff
@@ -466,6 +517,51 @@ def _pad(self, x, env_ids):
         return ((self._current_pos[env_ids] - x - 1) //
                 self._max_length) * self._max_length + x
 
+    def _set_default_return(self, env_ids):
+        mask = torch.ones_like(env_ids, dtype=torch.bool)
+        if torch.any(mask):
+            current_pos = self._current_pos[env_ids]
+            current_pos -= 1
+            ind = (env_ids[mask], self.circular(current_pos[mask]))
```
Comment (emailweixu) [inline alf/experience_replayers/replay_buffer.py]: Isn't env_ids[mask] same as env_ids since mask is an all-one tensor?
Same question for current_pos[mask]

File: alf/experience_replayers/replay_buffer.py [hunk 8]
```diff
@@ -466,6 +517,51 @@ def _pad(self, x, env_ids):
         return ((self._current_pos[env_ids] - x - 1) //
                 self._max_length) * self._max_length + x
 
+    def _set_default_return(self, env_ids):
+        mask = torch.ones_like(env_ids, dtype=torch.bool)
+        if torch.any(mask):
```
Comment (emailweixu) [inline alf/experience_replayers/replay_buffer.py]: Basically, this condition is True if env_ids is not empty.

File: alf/experience_replayers/replay_buffer.py [hunk 9]
```diff
@@ -90,6 +99,23 @@ def __init__(self,
...
+                NOTE:
+                1) Reward transformations like ``RewardClipping.minmax=(-1, 1)'' set in
+                    ``TrainerConfig.data_transformer_ctor'' have to be set manually for
+                    the ReplayBuffer to be consistent:
+                    ``ReplayBuffer.reward_clip=(-1,1)''.
+                2) Discount ``gamma'' needs to be set consistent with ``TDLoss.gamma''.
+                3) Assumes ``keep_episodic_info'' to be True.
+            default_return (float): The default values of ``discounted_return''
+                when the episode has not ended.  It should be a very small value for
+                value target lower bounding.
```
Comment (emailweixu) [inline alf/experience_replayers/replay_buffer.py]: Comment that its value should not be bigger than the smallest possible discounted return.
Also note that it can be 0 if reward is always non-negative.

File: alf/experience_replayers/replay_buffer.py [hunk 10]
```diff
@@ -466,6 +517,51 @@ def _pad(self, x, env_ids):
...
+            ind = (env_ids[mask], self.circular(current_pos[mask]))
+            self._episodic_discounted_return[ind] = self._default_return
+
+    def _compute_store_episodic_return(self, env_ids, discount):
+        # Always pass in env_ids whose step_type is LAST, to save computation.
+        # overall mask
+        mask = torch.ones_like(env_ids, dtype=torch.bool)
+        # mask that shrinks as we are done with some env_ids
+        discounted_return = torch.zeros_like(env_ids, dtype=torch.float32)
+        if torch.any(mask):
```
Comment (emailweixu) [inline alf/experience_replayers/replay_buffer.py]: this is always true if env_ids is not empty

File: alf/experience_replayers/replay_buffer.py [hunk 11]
```diff
@@ -466,6 +517,51 @@ def _pad(self, x, env_ids):
...
+            mask &= discount == 0
+            env_ids = env_ids[mask]
+            current_pos = self._current_pos[env_ids]
+            current_pos -= 1
+            # Only store episodic return when game ends, not ending with timelimit.
```
Comment (emailweixu) [inline alf/experience_replayers/replay_buffer.py]: We may consider bootstrapping the the episodic return with the default_return if game ends with timelimit. For example, the episode length of Atari breakout is limited to 4500, which can be reached quite often.

File: alf/experience_replayers/replay_buffer.py [hunk 12]
```diff
@@ -466,6 +517,51 @@ def _pad(self, x, env_ids):
...
+        while torch.any(mask):
```
Comment (emailweixu) [inline alf/experience_replayers/replay_buffer.py]: When num_parallel_environments is large, this can cost a lot of time since there will be lots of game endings.
It is possible to do this without using loop, similar to what MCTSAlgorithm._backup() does.

File: alf/algorithms/agent.py (hunk 6 above)
Comment (le-horizon) [reply alf/algorithms/agent.py]: There will be more to be passed, such as goal_distance in HER relabeling.

File: alf/experience_replayers/replay_buffer.py (hunk 9 above)
Comment (le-horizon) [reply alf/experience_replayers/replay_buffer.py]: Done

File: alf/experience_replayers/replay_buffer.py (hunk 8 above)
Comment (le-horizon) [reply alf/experience_replayers/replay_buffer.py]: Removed

File: alf/experience_replayers/replay_buffer.py (hunk 7 above)
Comment (le-horizon) [reply alf/experience_replayers/replay_buffer.py]: Removed.  Most likely copy pasted but forgot to remove.

File: alf/experience_replayers/replay_buffer.py (hunk 10 above)
Comment (le-horizon) [reply alf/experience_replayers/replay_buffer.py]: Removed

File: alf/experience_replayers/replay_buffer.py (hunk 11 above)
Comment (le-horizon) [reply alf/experience_replayers/replay_buffer.py]: Commented.

File: alf/experience_replayers/replay_buffer.py (hunk 12 above)
Comment (le-horizon) [reply alf/experience_replayers/replay_buffer.py]: Good point.  Done.

**Review by le-horizon — COMMENTED** [review]
@emailweixu Addressed all comments, could you take another look?

Thanks,
Le

File: alf/experience_replayers/replay_buffer.py [hunk 13]
```diff
@@ -518,49 +523,50 @@ def _pad(self, x, env_ids):
...
+                     **torch.arange(max_len, dtype=torch.float32)).unsqueeze(0)
+        epi_discounts = discounts * self._buffer.discount[all_ind]
+        # rewards = [r0, r1, r2, r3]
+        rewards = self._buffer.reward[all_ind]
+        rewards[~valid] = 0
+        epi_discounts[~valid] = 0
+        # reward[t] * discount[t - 1] is the correct discounted_return.
+        discounted_rewards = rewards[:, 1:] * epi_discounts[:, :-1]
+        # Then flip and cumsum from back of episode.
+        discounted_return = discounted_rewards.flip(1).cumsum(dim=1).flip(1)
```
Comment (emailweixu) [inline alf/experience_replayers/replay_buffer.py:582]: There is a problem with this if some TimeStep.discount is 0 in the middle of the episode. cumsum will add the reward after the zero discount towards the steps before it.
Is this tested in unittest?

File: alf/experience_replayers/replay_buffer.py [hunk 14]
```diff
@@ -518,49 +523,50 @@ def _pad(self, x, env_ids):
...
+        future_discounted_return = discounted_return / discounts[:, :-1]
+        valid_backfill = valid[:, :-1]
+        valid_ind = (env_ids.unsqueeze(1).expand(-1,
+                                                 max_len - 1)[valid_backfill],
+                     self.circular(epi_pos[:, :-1]).expand(
```
Comment (emailweixu) [inline alf/experience_replayers/replay_buffer.py]: `epi_pos[:, :-1]` already has the right shape, no need to do `expand()`. And `circular(epi_pos)` was already calculated at line 546

File: alf/algorithms/agent.py (hunk 6 above)
Comment (emailweixu) [reply alf/algorithms/agent.py]: Isn't HER relabeling a DataTransformer? It can populate rollout_info by itself.

File: alf/experience_replayers/replay_buffer.py (hunk 13 above)
Comment (le-horizon) [reply alf/experience_replayers/replay_buffer.py:582]: Fixed and added tests.

**Review by le-horizon — COMMENTED** [review]
Addressed all comments, PTAL @emailweixu 

Thanks

File: alf/experience_replayers/replay_buffer.py [hunk 15]
```diff
@@ -90,6 +99,24 @@ def __init__(self,
             allow_multiprocess (bool): whether multiprocessing is supported.
             keep_episodic_info (bool): index episode start and ending positions.
                 If None, its value will be set to True if ``num_earliest_frames_ignored``>0
+            record_episodic_return (bool): If True, computes and stores episodic return for
+                every step in the episode upon episode completion.  The field
+                ``discounted_return'' stores the information.  When episodes are
```
Comment (Haichao-Zhang) [inline alf/experience_replayers/replay_buffer.py]: '' -> ``. This is a minor issue that appeared multiple times in this section.

File: alf/experience_replayers/replay_buffer.py [hunk 16]
```diff
@@ -549,6 +694,17 @@ def get_episode_begin_position(self, pos, env_ids):
...
 
+    def get_disc_0_begin_position(self, pos, env_ids):
+        """
+        Note that the discount 0 step may no longer be in the replay buffer.
+        """
+        indices = (env_ids, self.circular(pos))
+        first_step_pos = self._prev_disc_0_pos[indices]
+        # buffer_step_types = self._buffer.step_type
+        # is_first = buffer_step_types[indices] == ds.StepType.FIRST
+        # first_step_pos[is_first] = pos[is_first]
```
Comment (Haichao-Zhang) [inline alf/experience_replayers/replay_buffer.py]: these lines can be removed now?

File: alf/test/main.py [hunk 17]
```diff
@@ -16,6 +16,8 @@
 import torch
 import unittest
 
+import alf
+
```
Comment (Haichao-Zhang) [inline alf/test/main.py]: These two lines can be removed?

File: alf/experience_replayers/replay_buffer.py (hunk 15 above)
Comment (le-horizon) [reply alf/experience_replayers/replay_buffer.py]: Ah, good catch.  Got mixed up with LaTex.

File: alf/experience_replayers/replay_buffer.py (hunk 16 above)
Comment (le-horizon) [reply alf/experience_replayers/replay_buffer.py]: Done

File: alf/test/main.py (hunk 17 above)
Comment (le-horizon) [reply alf/test/main.py]: Good catch.  It was from a git pick from an outdated branch..

**Review by le-horizon — COMMENTED** [review]
Done all, thanks for the careful review Haichao.

File: alf/experience_replayers/replay_buffer.py [hunk 18]
```diff
@@ -282,15 +330,32 @@ def add_batch(self, batch, env_ids=None, blocking=False):
...
-                self._store_episode_end_pos(non_first, overwriting_pos,
+                self._store_episode_end_pos(~epi_first, overwriting_pos,
                                             env_ids)
                 # 3.3. initialize episode beginning positions to itself
-                epi_first, = torch.where(step_types == ds.StepType.FIRST)
                 self._indexed_pos[(env_ids[epi_first],
                                    self.circular(overwriting_pos[epi_first])
                                    )] = overwriting_pos[epi_first]
+                if self._record_episodic_return:
+                    f_0 = (batch.discount == 0) | (
```
Comment (emailweixu) [inline alf/experience_replayers/replay_buffer.py]: `f_0 = (batch.discount == 0) |  epi_first`

File: alf/experience_replayers/replay_buffer.py [hunk 19]
```diff
@@ -535,7 +549,15 @@ def _compute_store_episodic_return(self, env_ids):
         # TODO: bootstrapping the the episodic return with the default_return
         # if game ends with timelimit. For example, the episode length of
         # Atari games is limited to 4500, which can be reached quite often.
-        first_step_pos = self.get_episode_begin_position(current_pos, env_ids)
+
+        # This is for computing from episode start, which shouldn't be used
+        # when computing episodic return every time discount = 0 step is added:
+        # first_step_pos = self.get_episode_begin_position(current_pos, env_ids)
```
Comment (emailweixu) [inline alf/experience_replayers/replay_buffer.py]: Is this comment still valid?

File: alf/experience_replayers/replay_buffer.py (hunk 18 above)
Comment (le-horizon) [reply alf/experience_replayers/replay_buffer.py]: Done

File: alf/experience_replayers/replay_buffer.py (hunk 19 above)
Comment (le-horizon) [reply alf/experience_replayers/replay_buffer.py]: It's valid.  Let me remove this since it's no longer used.

**Review by le-horizon — COMMENTED** [review]
Done all.  PTAL.

Thanks,
Le

**Review by Haichao-Zhang — APPROVED** [review]
LGTM

**Review by emailweixu — APPROVED** [review]

File: alf/algorithms/td_loss.py [hunk 20]
```diff
@@ -146,6 +146,18 @@ def compute_td_target(self, info: namedtuple, target_value: torch.Tensor):
                 td_lambda=self._lambda)
             returns = advantages + target_value[:-1]
 
+        disc_ret = ()
+        if hasattr(info, "discounted_return"):
+            disc_ret = info.discounted_return
+        if disc_ret != ():
+            with alf.summary.scope(self._name):
+                episode_ended = disc_ret > self._default_return
```
Comment (hnyu) [inline alf/algorithms/td_loss.py:154]: Where is this `self._default_return` set? By default it's not in `__init__`. If it's only set in a special case, then I think this discounted return thing shouldn't appear here in a general `TDLoss`. Maybe should drive a subclass from `TDLoss` for your purpose, or put this summary code somewhere else outside `TDLoss`.

Comment (le-horizon) [reply alf/algorithms/td_loss.py:154]: Good catch.  PR#1311 will fix this, and also create LowerBoundedTDLoss as a subclass of TDLoss.
https://github.com/HorizonRobotics/alf/pull/1311/files/87b86c1a819136137ddd13bf945af8e2e9dad23b#diff-615ff6b0217f23a46d9383ae639dafa53d3ef7a47c2266d10d3656f3fd1865f3
